src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,118 +2501,328 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+                }
+
+                var diagnostics = DiagnosticBag.GetInstance();
+                declaredMembersAndInitializers = buildDeclaredMembersAndInitializers(diagnostics);
+
+                var alreadyKnown = Interlocked.CompareExchange(ref _lazyDeclaredMembersAndInitializers, declaredMembersAndInitializers, DeclaredMembersAndInitializers.UninitializedSentinel);
+                if (alreadyKnown != DeclaredMembersAndInitializers.UninitializedSentinel)
+                {
+                    diagnostics.Free();
+                    return alreadyKnown;
+                }
+
+                AddDeclarationDiagnostics(diagnostics);
+                diagnostics.Free();
+
+                return declaredMembersAndInitializers!;
+            }
+
+            // Builds explicitly declared members (as opposed to synthesized members).
+            // This should not attempt to bind any method parameters as that would cause
+            // the members being built to be captured in the binder cache before the final
+            // list of members is determined.
+            DeclaredMembersAndInitializers? buildDeclaredMembersAndInitializers(DiagnosticBag diagnostics)
+            {
+                var builder = new DeclaredMembersAndInitializersBuilder();
+                AddDeclaredNontypeMembers(builder, diagnostics);
+
+                switch (TypeKind)
+                {
+                    case TypeKind.Struct:
+                        CheckForStructBadInitializers(builder, diagnostics);
+                        CheckForStructDefaultConstructors(builder.NonTypeNonIndexerMembers, isEnum: false, diagnostics: diagnostics);
+                        break;
+
+                    case TypeKind.Enum:
+                        CheckForStructDefaultConstructors(builder.NonTypeNonIndexerMembers, isEnum: true, diagnostics: diagnostics);
+                        break;
+
+                    case TypeKind.Class:
+                    case TypeKind.Interface:
+                    case TypeKind.Submission:
+                        // No additional checking required.
+                        break;
+
+                    default:
+                        break;
+                }
+
+                if (IsTupleType)
+                {
+                    builder.AddOrWrapTupleMembers(this);
+                }
+
+                // We already built the explicitly declared members and initializers on another thread, we might have detected that condition
+                // during member building on this thread and bailed, which results in incomplete data in the builder.
+                // In such case we have to avoid creating the instance of MemberAndInitializers since it checks the consistency
+                // of the data in the builder and would fail in an assertion if we tried to construct it from incomplete builder.
+                if (Volatile.Read(ref _lazyDeclaredMembersAndInitializers) != DeclaredMembersAndInitializers.UninitializedSentinel)
```

Comment:
Do we need to use `Volatile.Read()` in these methods? (It seems difficult to ensure we're doing this correctly.) If all of the reads and writes to the `_lazy` fields were in the top-level method (and either the local functions were inlined or the local functions simply calculated and returned values rather than setting fields), could we remove the `Volatile.Read()` calls?

Reply:
The volatile read not necessary, but gives us slightly better chance of avoiding doing some work.

---
In reply to: [564121130](https://github.com/dotnet/roslyn/pull/50650#discussion_r564121130) [](ancestors = 564121130)